### PR TITLE
Add Random Snow

### LINKED ISSUE #357: Add RandomSnow preprocessing layer
## Weather Augmentation

One of the real-world scenarios that pose challenges for training neural networks of Autonomous vehicles

output_20_0

Ref. Imple.

- https://albumentations.ai/docs/api_reference/augmentations/transforms/#albumentations.augmentations.transforms.RandomSnow
- https://github.com/UjjwalSaxena/Automold--Road-Augmentation-Library

### PR DESCRIPTION
**Pull request by kartik4949**
closes
#357 

Details:

## Original implementation requires

RGB to HLS conversion

H - hue
L - lightness
S - saturation

Original implementation scale `L` channel with appropriate value on some defined points

#### Im converting RGB to HSV

H - hue
S - sat
V - Value / Brightness

calculating `L` lightness in HSL space
with formula (min(RGB) + max(RGB) ) / 2

Scaling `L` on snow points with corresponding scalar (i.e brightness coefficient)

Now, 
there is not function as HSL to RGB in tensorflow hence
need to get corresponding 

`S` and `V` values from the new scaled `L` by using some maths.
@LukeWood please check it, need your review!

official formulas:
https://en.wikipedia.org/wiki/HSL_and_HSV#HSL_to_RGB
image

### CODE REVIEW

**Comment by kartik4949** [comment]
results:
snow2
snow3

**Comment by kartik4949** [comment]
@bhack can u point to rgb_to_hsv code?

**Comment by kartik4949** [comment]
We have one major issue though
HSV to RGB is not a continuous function hence gradient is not implemented/buggy in tensorflow!!

**Comment by kartik4949** [comment]
Also,
In HSV, V corresponds to brightness or how much light thrown on the color.

hence increasing this value might give snow like effect but it won't entirely make white snow, the underlying color effect will be there:(

**Comment by kartik4949** [comment]
@bhack @LukeWood should I close this, I believe its not relevant now, as 
experimental weather KPL are not academically test well.

**Comment by LukeWood** [comment]
> @bhack @LukeWood should I close this, I believe its not relevant now, as experimental weather KPL are not academically test well.

Why don't you leave it open for now and I can ask some self driving CV engineers I know if they would use these weather KPLs or if they are too untested.  Its interesting no matter what.

**Comment by bhack** [comment]
We have already some early result for the snow leaderboards:

https://acdc.vision.ee.ethz.ch/benchmarks

But probably we will need to wait for the workshop next month to check the papers details.

**Comment by kartik4949** [comment]
> We have already some early result for the snow leaderboards:
> 
> https://acdc.vision.ee.ethz.ch/benchmarks
> 
> But probably we will need to wait for the workshop next month to check the papers details.

This is useful!

**Comment by LukeWood** [comment]
closed due to inactivity, feel free to re-open if you re-implement w/ Tensorflow ops.
